src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs

```diff
@@ -270,39 +270,39 @@ public virtual string CreateToken(SecurityTokenDescriptor tokenDescriptor)
             if (tokenDescriptor.Audience != null)
             {
                 if (payload.ContainsKey(JwtRegisteredClaimNames.Aud))
-                    LogHelper.LogInformation(LogHelper.FormatInvariant(LogMessages.IDX14113, nameof(tokenDescriptor.Audience)));
+                    LogHelper.LogInformation(LogHelper.FormatInvariant(LogMessages.IDX14113, LogHelper.MarkAsNonPII(nameof(tokenDescriptor.Audience))));
```

Comment:
There are a couple of things here that could be considered, some were here already.
1. We don't know if we are going to log this message yet, it depends on if the EventSource is setup and the level. Yet we are creating and formatting a string.
2. We are creating a new object on the stack with LogHelper.MarkAsNonPII creates a new object.

Could we delay the action of creating a string and the new object until we know we will be logging?
It is not as important when we are throwing, as that interrupts the stack and we assume this is infrequent.
This work will occur in the positive case.

Reply:
@sruke, @brentschmaltz 
I was wondering, if we couldn't step back and encode the non-PII in the log message. For instance, we could have a little sign rarely used in messages (`~`) added to the log messages (before or after the {0} {1}). This would tag that this data is not PII

This way this would be data driven (contained in the log message) and auto descriptive, and we would not need to do any changes in the line of code calling LogHelper.LogXX(). This would be by the LogHelper.LogXX() methods themselves

Of course LogHelper.LogXX() would apply or not the PII to each argument depending on the little sign, and strip away the little sign

For instance:

```CSharp
internal const string IDX14113 = "IDX14113: A duplicate value for 'SecurityTokenDescriptor.{0}~' exists in 'SecurityTokenDescriptor.Claims'. \nThe value of 'SecurityTokenDescriptor.{0}~' is used.";
```

would contain the `~` sign because there is no PII. No extra work, and no extra allocation here
It would be called as expected:

```CSharp
LogHelper.LogInformation(LogHelper.FormatInvariant(LogMessages.IDX14113, nameof(tokenDescriptor.Audience)));
```

Another advantage is we could automatically publish a doc of the log messages (request from office of the CTO) which would also explain what is PII or not ...

(sorry for jumping late on this, I realize it's a different design)

Reply:
That's a really interesting idea @jmprieur and @brentschmaltz had some good comments too. However, this would require us to revisit the complete design on when to categorize arguments into PII and NonPII. 

Current design considers all arguments as PII and scrubs them before the EventLevel for logging is decided. We made a design decision to intercept this and mark system metadata as NonPII that required us to wrap individual arguments as NonPII and create new objects. A similar feature was introduced in SAL recently. To be consistent shall we proceed with what we have?

Reply:
I realize this is a completely different design, but this might be worth discussing? (again, sorry for being late on this).

Reply:
@sruke @jmprieur that little '~' would address my concerns nicely, of course we need to be sure the ~ is from us and not an attacker. It is nice because we don't need the PII wrapper.

This idea made me think of an approach would be to use params where the first value defines the number of PII values.
For example:
params: 1,foo, bar would tell us the first string (foo) is PII.
this would force us to rearrange the strings in many of our messages.

I am not too concerned if we have to revisit some previous work as, finding the right solution that is natural and performant sometimes takes a couple of tries.

My main concern is that we create strings and objects that will never be logged in the success path.

Reply:
+1 : create only when needed and use templating where possible so the strings are max created once. I suppose that will be addressed in the next pass of the string creation. 

@brentschmaltz wondering if any customers are currently depending on any of the strings and may see a re-ordering as a "breaking" change. hopefully they only depend on the id and not the actual string, so we can optimize how we generate the strings.

Reply:
@henrik-me I should clarify the 'reorder' idea.
What i mean is if a string is $"foo {0}, {1}, {2}" with parameters x, z, y and 0 and 2 are non pii, then we would have to modify the string to be
$"foo {0}, {2}, {1}" with parameters x, y, z

the user would see the same string.

Reply:
I will merge this PR into Wilson as it has some worthwhile work that is helpful to the community and come up with a one pager summarizing your proposals and how their potentials implementations would look like and we can have another discussion with the main goal to limit unnecessary string generations. Once we have a path forward I'll revisit this feature.